File: server/src/engine/EffectResolver.ts
/**
 * EffectResolver (IP-4010) — engagement gating, structural Deceive/Destroy dispatch, cumulative
 * Degrade stacking, and the King denial-streak tracker (BL-0015: stored field, not recomputed).
 */
import type {
  Asset,
  EffectStateEntry,
  EngagementResult,
  FiveDsEffect,
  OrbitalRegimeLabel,
  PlayerState,
} from '@owchess/shared';
import type { BeliefState } from './BeliefState.js';

/** FS-105's pinned numeric refinement: durations short enough that no single application wins
 *  outright against the 6-turn mission-denial threshold (FR-4400). */
const DISRUPT_DENY_DURATION = 3;
const DEGRADE_DURATION = 4;

export class EffectResolver {
  /**
   * FR-4002: requires at least 'target'-level precision (read from the effector's own player's
   * beliefOfOpponent — a deviation matching BL-0028/BL-0033's pattern, since GDS-09's signature
   * has no state-access parameter either; see this package's Deviation note). Destroy flags the
   * target destroyed; Deceive calls BeliefState.applyDeception instead of mutating target's own
   * true state (the structural distinction GDS-04/07 require, enforced here, not merely
   * described).
   */
  resolveEngagement(
    effectorObserverState: PlayerState,
    effector: Asset,
    target: Asset,
    effect: FiveDsEffect,
    beliefState: BeliefState,
    currentTurn: number,
    falseRegime?: OrbitalRegimeLabel
  ): EngagementResult {
    const belief = effectorObserverState.beliefOfOpponent.get(target.assetId);
    if (belief?.precision !== 'target') {
      return { effect, success: false, reason: 'insufficient targeting-quality data (FR-4002)' };
    }

    switch (effect) {
      case 'destroy':
        target.destroyed = true;
        return { effect, success: true };

      case 'deceive':
        beliefState.applyDeception(
          effectorObserverState,
          target.assetId,
          falseRegime ?? target.trueRegime,
          currentTurn,
          effector.assetId
        );
        return { effect, success: true };

      case 'disrupt':
      case 'deny':
        this.addEffectEntry(target, effect, DISRUPT_DENY_DURATION, currentTurn, effector.assetId);
        return { effect, success: true };

      case 'degrade':
        this.addEffectEntry(target, 'degrade', DEGRADE_DURATION, currentTurn, effector.assetId);
        return { effect, success: true };

      default:
        return { effect, success: false, reason: `unknown effect kind ${effect}` };
    }
  }

  private addEffectEntry(
    target: Asset,
    kind: 'disrupt' | 'deny' | 'degrade',
    durationTurns: number,
    currentTurn: number,
    sourceEffectorAssetId: string
  ): void {
    // FR-4300: multiple concurrent Degrade entries coexist and tick independently, they don't
    // overwrite. A fresh Disrupt/Deny application is likewise its own new entry.
    const entry: EffectStateEntry = {
      kind,
      appliedTurn: currentTurn,
      durationTurns,
      stackCount: 1,
      sourceEffectorAssetId,
    };
    target.activeEffects.push(entry);
  }

  /**
   * Decrements every active effect's remaining duration; expires (removes) entries that reach
   * zero. Updates the King's consecutiveDenialTurns (BL-0015): increments while >=1 qualifying
   * Disrupt/Deny/Degrade entry is active, resets to 0 the instant none are.
   */
  tickActiveEffects(asset: Asset, currentTurn: number): void {
    asset.activeEffects = asset.activeEffects.filter((entry) => {
      if (entry.durationTurns === 'until-cleared') return true;
      const elapsed = currentTurn - entry.appliedTurn;
      return elapsed < entry.durationTurns;
    });

    if (asset.isKing) {
      const denied = asset.activeEffects.some(
        (e) => e.kind === 'disrupt' || e.kind === 'deny' || e.kind === 'degrade'
      );
      if (denied) {
        asset.consecutiveDenialTurns += 1;
        asset.totalDenialTurns += 1;
      } else {
        asset.consecutiveDenialTurns = 0;
      }
    }
  }
}

File: python-mvp/engine/effect_resolver.py
"""
EffectResolver — applies a Five D's effect to a target's true state on successful engagement,
gated by 'target'-level belief precision. Mirrors server/src/engine/EffectResolver.ts.
"""
from __future__ import annotations

from typing import Optional

from .belief_state import BeliefState
from .types import Asset, EffectStateEntry, FiveDsEffect, OrbitalRegimeLabel, PlayerState

DISRUPT_DENY_DURATION = 3
DEGRADE_DURATION = 4


class EffectResolver:
    def resolve_engagement(
        self,
        effector_observer_state: PlayerState,
        effector: Asset,
        target: Asset,
        target_owner_state: PlayerState,
        effect: FiveDsEffect,
        belief_state: BeliefState,
        current_turn: int,
        false_regime: Optional[OrbitalRegimeLabel] = None,
    ) -> tuple[bool, Optional[str]]:
        belief = effector_observer_state.belief_of_opponent.get(target.asset_id)
        if belief is None or belief.precision != "target":
            return False, "insufficient targeting-quality data (need 'target'-level precision)"

        if effect == "destroy":
            target.destroyed = True
            return True, None

        if effect == "deceive":
            # Sprint-0 fix: plants a false regime in the *target's owner's* belief about the
            # effector's own asset — see belief_state.py's module docstring for the reasoning.
            resolved_false_regime = false_regime or effector.true_regime
            belief_state.apply_deception(
                target_owner_state, effector.asset_id, resolved_false_regime, current_turn, target.asset_id
            )
            return True, None

        if effect in ("disrupt", "deny"):
            self._add_effect_entry(target, effect, DISRUPT_DENY_DURATION, current_turn, effector.asset_id)
            return True, None

        if effect == "degrade":
            self._add_effect_entry(target, "degrade", DEGRADE_DURATION, current_turn, effector.asset_id)
            return True, None

        return False, f"unknown effect kind {effect}"

    def _add_effect_entry(
        self, target: Asset, kind: str, duration_turns: int, current_turn: int, source_effector_asset_id: str
    ) -> None:
        target.active_effects.append(
            EffectStateEntry(
                kind=kind,  # type: ignore[arg-type]
                applied_turn=current_turn,
                duration_turns=duration_turns,
                stack_count=1,
                source_effector_asset_id=source_effector_asset_id,
            )
        )

    def tick_active_effects(self, asset: Asset, current_turn: int) -> None:
        kept = []
        for entry in asset.active_effects:
            if entry.duration_turns == "until-cleared":
                kept.append(entry)
                continue
            elapsed = current_turn - entry.applied_turn
            if elapsed < entry.duration_turns:
                kept.append(entry)
        asset.active_effects = kept

        if asset.is_king:
            denied = any(e.kind in ("disrupt", "deny", "degrade") for e in asset.active_effects)
            if denied:
                asset.consecutive_denial_turns += 1
                asset.total_denial_turns += 1
            else:
                asset.consecutive_denial_turns = 0
